icu4c/source/test/intltest/loctest.cpp

```diff
@@ -3341,3 +3393,117 @@ void LocaleTest::TestBug11053UnderlineTimeZone() {
     Locale l8(locale_str.c_str());
     assertTrue((locale_str + " !l8.isBogus()").c_str(), !l8.isBogus());
 }
+
+void LocaleTest::TestUnd() {
+    IcuTestErrorCode status(*this, "TestUnd()");
+
+    static const char empty[] = "";
+    static const char root[] = "root";
+    static const char und[] = "und";
+
+    Locale empty_ctor(empty);
+    Locale empty_tag = Locale::forLanguageTag(empty, status);
+    status.errIfFailureAndReset("\"%s\"", empty);
+
+    Locale root_ctor(root);
+    Locale root_tag = Locale::forLanguageTag(root, status);
+    status.errIfFailureAndReset("\"%s\"", root);
+
+    Locale und_ctor(und);
+    Locale und_tag = Locale::forLanguageTag(und, status);
+    status.errIfFailureAndReset("\"%s\"", und);
+
+    assertEquals("getName()", empty, empty_ctor.getName());
+    assertEquals("getName()", empty, root_ctor.getName());
```

Comment:
Different from what we decided: Mapping "root" to "" without canonicalizing.
I don't remember why we decided to only do it in canonicalize().
Replying to your email thread with details.

Reply:
Decided via email that in fact we want this behavior of mapping both "und" and "root" to "" right in the constructor.